src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/NullableConverter.cs

```diff
@@ -27,13 +30,16 @@ internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert,
                 return true;
             }
 
-            state.Current.JsonPropertyInfo = state.Current.JsonTypeInfo.ElementTypeInfo!.PropertyInfoForTypeInfo;
-            if (_elementConverter.TryRead(ref reader, typeof(T), options, ref state, out T element))
+            JsonTypeInfo previousTypeInfo = state.Current.JsonTypeInfo;
+            state.Current.JsonTypeInfo = state.Current.JsonTypeInfo.ElementTypeInfo!;
+            if (_elementConverter.OnTryRead(ref reader, typeof(T), options, ref state, out T element))
             {
                 value = element;
+                state.Current.JsonTypeInfo = previousTypeInfo;
                 return true;
             }
 
+            state.Current.JsonTypeInfo = previousTypeInfo;
```

Comment:
Why do we now need to roll back the JsonTypeInfo after serialization is done?

Reply:
Short answer is that I got more than 100 test failures when I don't do it. Longer answer:
Many asserts validating JsonTypeInfo matches after the operation - technically asserts can be removed since they're not valid if value is not used later (or special cased for nullable) because most of the time that is fine but not in some cases: continuation + collection + nullable we sometimes need the correct type info. When all asserts all disabled (i.e. Release) I'm getting null reference exceptions on dictionaries trying to access element type but converter is nullable, it happens when we need to handle continuation correctly (`Pop(false)`)